topics/products/phpstorm/php_open_api_php_type_providers.md

```diff
@@ -0,0 +1,304 @@
+
+Suppose we have two files:
+
+<path>foo.php</path>:
+
+```
+<?php
+
+function foo(): string {
+  return "Hello World!";
+}
+```
+
+<path>main.php</path>:
+
+```
+<?php
+
+$a = foo();
+echo $a;
+```
+
+In the `main.php` file, we call the `foo()`, which is defined in another `foo.php` file.
+Because of this, PhpStorm won't be able to infer the type of the `$a` variable during the indexing stage, since it depends on the definition of the `foo()` from another file.
+
+For such cases, PhpStorm will create an Incomplete type in which writes all the necessary information to resolve the type when indexing is finished.
+In this case, it's the name of the function being called, so PhpStorm will create an Incomplete type `#F\foo`.
+
+#### Incomplete Types Structure
+
+At the beginning there is a `#` character, which is a marker that the type is Incomplete.
+
+It's followed by the type provider's unique key.
+PhpStorm uses this key to decide which type provider to pass the Incomplete type to resolve.
+
+The rest of the line is the encoded information.
+In the example above, this is the fully qualified name of the function.
+
+PhpStorm will further pass the created Incomplete type to the `PhpTypeProvider4.complete()` after indexing is finished to resolve it into a Complete type.
+
+During indexing, PhpStorm collects information about all types of elements in this way and stores them in the index.
+When there is a need for the type of some expression, PhpStorm passes the Incomplete type obtained at the indexing stage for resolving.
+
+#### Incomplete Types Resolving
+
+The second phase of type inference is the global Incomplete type resolution.
+At this phase, PhpStorm calls `PhpTypeProvider4.complete()` of each type provider.
+All Incomplete types are passed to the providers that created them.
+At this point, PhpStorm can access any information from other files to resolve the Incomplete type.
+
+### Union Types
+
+Since PHP is a dynamically typed language, at the type inference stage PhpStorm can get a situation where the type can be either one or the other.
+
+For example:
+
+```
+<?php
+
+$a = 100;
+if (rand(0, 10) > 5) {
+  $a = "Hello World!";
+}
+
+$a; // (1)
+```
+
+In (1), the variable `$a` will be of type `int|string` because PhpStorm can't deduce exactly which branch the execution will take.
+
+PhpStorm stores this types as separate strings inside the [`PhpType`](#phptype) class.
+Each of the types can be either Complete or Incomplete.
+In the Incomplete type resolving process, PhpStorm will resolve each union type individually.
+
+Since some providers may return types for the same PSI element, union types may appear for some elements.
+
+## PhpType
+
+PhpStorm uses the `com.jetbrains.php.lang.psi.resolve.types.PhpType` class to work with types.
+
+To add types to it, use `add()`, which can take either another `PhpType` or a `string`.
+
+To check that a type is Complete, use `isComplete()`.
+
+To resolve the Incomplete type, use `global()`.
```

Comment:
It's worth to repeat the context IMO, please add

"This method shouldn't be used during indexing, namely inside `PhpTypeProvider4#getType`"

Reply:
Fixed.